### PR TITLE
Add a helper TinyStr crate

### PR DESCRIPTION
**Pull request by zbraniecki**
I separated out the new helper crate from #7, applied @raphlinus feedback and added benchmarks.

### CODE REVIEW

**Comment by zbraniecki** [comment]
@ralphinus - would you be willing to review this?

**Comment by Manishearth** [comment]
@zbraniecki I was suggesting this be a separate repo :smile:, I think this crate has utility outside of unic-locale

**Comment by Manishearth** [comment]
Also, crate name bikeshed: `filament`

(tinystr may be more searchable)

**Comment by zbraniecki** [comment]
Yep! I'd like to first get a review of this code, before I land it - but I'll land it in a separate repo.

> (tinystr may be more searchable)

Yeaaa... I prefer searchability. I also am considering `tinystring` because it allocates.

**Comment by Manishearth** [comment]
Wait, how does it allocate?

**Comment by zbraniecki** [comment]
> Wait, how does it allocate?

Does it not? It owns the data it stores, right?

**Comment by raphlinus** [comment]
It shouldn't allocate, it uses a scalar. So it can happily be #[no_std].

[Filament](https://github.com/google/filament) is a superb physically based rendering crate, but I understand the appeal.

**Comment by Manishearth** [comment]
> Does it not? It owns the data it stores, right?

It stores things inline in an integer

**Comment by zbraniecki** [comment]
Ah, ok. Seems like my understanding of terminology is fuzzy :( Sorry for that! `tinystr` is good.

I have one remaining open item from @raphlinus review in the previous issue:

```
(about the unchecked constructor) One thing to watch out for is endianness: I recommend the TinyStr constructor always take the string in little-endian, so do .from_le() to swap on a BE system. It's very difficult to test on big-endian systems, as they're rare (but do exist).
```

I'm not sure how to do that - I'm not familiar with BE/LE systems so I also don't know how to test it.

**Comment by raphlinus** [comment]
> I'm not sure how to do that - I'm not familiar with BE/LE systems so I also don't know how to test it.

I'm willing to desk-check, but here's the basic idea. The struct always stores in "native endianness" so that the transmute to `&str` is zero cost. Thus, "abcd" is 0x64636261 on LE systems and 0x61626364 on BE. What I'm recommending is that we always use LE in the API, so the constructor takes  0x64636261 in both cases, calls `.from_le()` on it which is a no-op on LE systems, but swaps bytes on BE.

A similar strategy is used to compute the various masks testing for letter classes, etc. This will have a small performance impact on BE systems, but makes it a lot easier to reason about.

**Comment by zbraniecki** [comment]
@raphlinus - thank you! Does it mean I should also do `to_le()` in `to_raw_parts()` ?

**Comment by raphlinus** [comment]
Yes, if (I haven't dug in) the intent is to provide a value that will be used by the constructor. The important thing is that `.to_le()` and `.from_le()` are balanced. For your macro case, it should all compile out even on BE machines, as these methods are `const fn`.

**Comment by zbraniecki** [comment]
Great! I added it to dumping and the unchecked constructor. I'll document it once we get to the docs phase.

@raphlinus - would you have time to skim through the code and verify if it looks like you intended it to look like? I'll then separate it to a new repo and try to maintain although a lot of the math in it is above my skill level :)

File: tinystr/src/helpers.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+use super::Error;
+
+pub(crate) unsafe fn make_4byte_str(
+    text: &str,
+    len: usize,
+    mask: u32,
+) -> Result<NonZeroU32, Error> {
+    // Mask is always supplied as little-endian.
+    let mask = mask.to_le();
```
Comment (raphlinus) [inline tinystr/src/helpers.rs]: An incredibly minor nit, but this should be `from_le`. The semantics are the same, but it communicates the intent of converting from a little-endian to a native-endian representation.

File: tinystr/src/tinystr4.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        ((!(lower + 0x1f1f_1f1f) | (lower + 0x0505_0505)) & mask) == 0
+    }
+
+    /// Makes the string all lowercase except for the first character,
+    /// which is made uppercase.
+    pub fn to_ascii_titlecase(self) -> TinyStr4 {
+        let word = self.0.get().to_le();
+        let mask = ((word + 0x3f3f_3f1f) & !(word + 0x2525_2505) & 0x8080_8080) >> 2;
+        let result = (word | mask) & !(0x20 & mask);
+        unsafe { TinyStr4(NonZeroU32::new_unchecked(result.to_le())) }
```
Comment (raphlinus) [inline tinystr/src/tinystr4.rs]: And this should be `from_le` as well, for the same reason.

File: tinystr/src/tinystr4.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                2 => make_4byte_str(text, 2, 0x8080).map(Self),
+                3 => make_4byte_str(text, 3, 0x0080_8080).map(Self),
+                4 => make_4byte_str(text, 4, 0x8080_8080).map(Self),
+                _ => Err(Error::InvalidSize),
+            }
+        }
+    }
+}
+
+impl Into<u32> for TinyStr4 {
```
Comment (raphlinus) [inline tinystr/src/tinystr4.rs]: It's unclear whether this is intended to convert into the native or portable (little-endian) form. Code as written is native.

File: tinystr/src/tinystr8.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        if len < 1 || len > 8 {
+            return Err(Error::InvalidSize);
+        }
+        unsafe {
+            let mut word: u64 = 0;
+            copy_nonoverlapping(text.as_ptr(), &mut word as *mut u64 as *mut u8, len);
+            let mask = 0x80808080_80808080u64 >> (8 * (8 - len));
+            // TODO: could do this with #cfg(target_endian), but this is clearer and
+            // more confidence-inspiring.
+            let mask = mask.to_le();
```
Comment (raphlinus) [inline tinystr/src/tinystr8.rs]: `from_le`

File: tinystr/src/tinystr8.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            if ((mask - word) & mask) != 0 {
+                return Err(Error::InvalidNull);
+            }
+            Ok(Self(NonZeroU64::new_unchecked(word)))
+        }
+    }
+}
+
+impl Into<u64> for TinyStr8 {
+    fn into(self) -> u64 {
```
Comment (raphlinus) [inline tinystr/src/tinystr8.rs]: Same as above.

**Comment by Manishearth** [comment]
Overall seems good to me aside from the endianness stuff

**Comment by zbraniecki** [comment]
Moved to a separate repo: https://github.com/zbraniecki/tinystr

**Pull request closed without merging**